### PR TITLE
NDB: Start in on `ndb.client.Client`

### PR DESCRIPTION
**Pull request by chrisrossi**
This is patterned off of `datastore.client.Client` and depends heavily on `google.cloud.client`.

### CODE REVIEW

File: ndb/src/google/cloud/ndb/client.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+"""A client for NDB which manages credentials, project, namespace."""
+
+import os
+
+from google.cloud import environment_vars
+from google.cloud.client import ClientWithProject
+from google.cloud import _helpers
+
+_DATASTORE_HOST = "datastore.googleapis.com"
```
Comment (dhermes) [inline ndb/src/google/cloud/ndb/client.py]: Why is this needed? Isn't there a constant in `google-cloud-datastore` we can use?

File: ndb/src/google/cloud/ndb/client.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    if project is None:
+        project = _get_gcd_project()
+
+    if project is None:
+        project = _helpers._determine_default_project(project=project)
+
+    return project
+
+
+class Client(ClientWithProject):
```
Comment (dhermes) [inline ndb/src/google/cloud/ndb/client.py]: Why not inherit from `google.cloud.datastore.client.Client`?

File: ndb/tests/conftest.py [hunk 3]
```diff
@@ -40,3 +43,28 @@ def reset_state():
     yield
     model.Property._FIND_METHODS_CACHE.clear()
     model.Model._kind_map.clear()
+
+
+@pytest.fixture
+def environ():
+    """Copy of ``os.environ``"""
+    original = os.environ.copy()
```
Comment (dhermes) [inline ndb/tests/conftest.py]: This copy isn't needed since you overwrite `os.environ`?

File: ndb/tests/conftest.py [hunk 4]
```diff
@@ -40,3 +43,28 @@ def reset_state():
...
+@pytest.fixture(autouse=True)
+def initialize_environment(request, environ):
+    """Set environment variables to default values.
+
+    There are some variables, like ``GOOGLE_APPLICATION_CREDENTIALS``, that we
+    want to reset for unit tests but not system tests. This fixture introspects
+    the current request, determines whether it's in a unit test, or not, and
+    does the right thing.
+    """
+    if request.module.__name__.startswith("tests.unit"):
```
Comment (dhermes) [inline ndb/tests/conftest.py]: TIL the [`request`][1] fixture, thanks!

[1]: https://docs.pytest.org/en/latest/reference.html#request

File: ndb/tests/unit/test_client.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    with patch:
+        yield creds
+
+
+class TestClient:
+    @staticmethod
+    def test_constructor_no_args():
+        with patch_credentials("testing"):
+            client = client_module.Client()
+        assert client.SCOPE == ("https://www.googleapis.com/auth/datastore",)
```
Comment (dhermes) [inline ndb/tests/unit/test_client.py]: Is this an instance attribute named `SCOPE`?

**Comment by chrisrossi** [comment]
> @chrisrossi I'm not clear why this special Client class is needed. A parallel to ndb would be https://github.com/googleapis/google-cloud-python-happybase, which doesn't define a Client of it's own (but does have a similar metaphor).

A couple of days ago, before starting this, I sent around an email entitled "NDB client story", where I was pretty much fishing for exactly this kind of input. Did you see that?

I can look at happybase--I'm not familiar with it at the moment. Basically, if we're going to interact with datastore, we need to get credentials, figure out the project/app, and possibly get a namespace. Since this is just a layer on datastore, I looked to datastore for the model on how to do it. I'm not wed to anything in particular, but I try not to invent new idioms if there is one already available, and I figured datastore probably represented some kind of "best practice" for Google APIs.

**Comment by chrisrossi** [comment]
> @dhermes
> 
> > @chrisrossi I'm not clear why this special Client class is needed. A parallel to ndb would be https://github.com/googleapis/google-cloud-python-happybase, which doesn't define a Client of it's own (but does have a similar metaphor).
> 
> A couple of days ago, before starting this, I sent around an email entitled "NDB client story", where I was pretty much fishing for exactly this kind of input. Did you see that?
> 
> I can look at happybase--I'm not familiar with it at the moment. Basically, if we're going to interact with datastore, we need to get credentials, figure out the project/app, and possibly get a namespace. Since this is just a layer on datastore, I looked to datastore for the model on how to do it. I'm not wed to anything in particular, but I try not to invent new idioms if there is one already available, and I figured datastore probably represented some kind of "best practice" for Google APIs.

I've looked at HappyBase this morning and it has a `Connection` object which is the moral equivalent of the `Client` object here. We need some way to manage the little bit of state that gets used to talk to Datastore. `google.cloud.datastore` uses a `Client` object and it makes sense to me for us also to use a `Client` object that is the parallel of that.

File: ndb/src/google/cloud/ndb/client.py (hunk 1 above)
Comment (chrisrossi) [reply ndb/src/google/cloud/ndb/client.py]: There is `google.cloud.datastore.client._DATASTORE_BASE_URL`, which could be used to derive the host. It's private, and you had mentioned copying private members from Datastore rather than depending on them. We don't use a URL, since we don't support the HTTP transport, but we could derive the host from the URL and may implement that HTTP transport in a future iteration. It's not clear to me there's a single right answer here.

File: ndb/src/google/cloud/ndb/client.py (hunk 2 above)
Comment (chrisrossi) [reply ndb/src/google/cloud/ndb/client.py]: Because `google.cloud.datastore.client.Client` just inherits from `ClientWithProject` and really only adds things that are specific to Datastore that would have to be overridden for NDB. I looked, and there didn't seem to be enough stuff we would want to reuse so I used it as a guide to follow instead.

File: ndb/tests/conftest.py (hunk 3 above)
Comment (chrisrossi) [reply ndb/tests/conftest.py]: raise FingersFasterThanBrainError()

Fixed: 78f02e7

File: ndb/tests/conftest.py (hunk 4 above)
Comment (chrisrossi) [reply ndb/tests/conftest.py]: :+1:

File: ndb/tests/unit/test_client.py (hunk 5 above)
Comment (chrisrossi) [reply ndb/tests/unit/test_client.py]: It's a class attribute. Ends up getting used by the super-class when authenticating, I believe.

File: ndb/src/google/cloud/ndb/client.py (hunk 2 above)
Comment (dhermes) [reply ndb/src/google/cloud/ndb/client.py]: OK

**Comment by dhermes** [comment]
@chrisrossi I'd say push on with `Client`, but try to think of a way **not** to have this if we can avoid it. `ndb`'s design is essentially going to force global state to be stored on the module, joy!

**Comment by chrisrossi** [comment]
> @chrisrossi I'd say push on with `Client`, but try to think of a way **not** to have this if we can avoid it. `ndb`'s design is essentially going to force global state to be stored on the module, joy!

@dhermes I'm reasonably confident we can contain all the state we need in `client.Client` and the thread locals in `_runstate`.

File: ndb/src/google/cloud/ndb/client.py (hunk 1 above)
Comment (dhermes) [reply ndb/src/google/cloud/ndb/client.py]: Isn't it defined anywhere in the `gapic` module(s)?

Comment (chrisrossi) [reply ndb/src/google/cloud/ndb/client.py]: Found it in _http. Which is technically still private, but I guess I'll use that.

**Comment by chrisrossi** [comment]
@dhermes Yep, fixed.

File: ndb/tests/conftest.py [hunk 6]
```diff
@@ -40,3 +43,28 @@ def reset_state():
     yield
     model.Property._FIND_METHODS_CACHE.clear()
     model.Model._kind_map.clear()
+
+
+@pytest.fixture
+def environ():
+    """Copy of ``os.environ``"""
+    original = os.environ
+    environ = original.copy()
```
[the remark this answers is not in the document]
Comment (chrisrossi) [reply ndb/tests/conftest.py]: Maybe not...

~~~~
(gcpdev) chris@opportunity:~/proj/google/clouddev/ndb$ pylint google.cloud.ndb tests | wc -l
2732
~~~~

[the remark this answers is not in the document]
Comment (chrisrossi) [reply ndb/tests/conftest.py]: The global is the function, not the import of os.environ, FWIW. Would be an actual problem if environ were a recursive function. "Fixed" anyway.

pylint also complains about line 30, because the arg `environ` masks the global `environ`, which is the fixture. Of course, the named argument, `environ`, is what signals to `pytest` to use the fixture named `environ`. 

Maybe we get bogged down in detail sometimes?
